pkg/maps/nodemap/node_map_v2.go

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package nodemap
+
+import (
+	"fmt"
+	"net"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/cilium/cilium/pkg/ebpf"
+)
+
+// compile time check of MapV2 interface
+var _ MapV2 = (*nodeMapV2)(nil)
+
+const (
+	MapNameV2 = "cilium_node_map_v2"
+)
+
+// Map provides access to the eBPF map node.
+type MapV2 interface {
+	// Update inserts or updates the node map object associated with the provided
+	// IP, node id, and SPI.
+	Update(ip net.IP, nodeID uint16, SPI uint8) error
+
+	// Delete deletes the node map object associated with the provided
+	// IP.
+	Delete(ip net.IP) error
+
+	// IterateWithCallback iterates through all the keys/values of a node map,
+	// passing each key/value pair to the cb callback.
+	IterateWithCallback(cb NodeIterateCallbackV2) error
+
+	// Size returns what how many entries the node map is configured to hold.
+	Size() uint32
+}
+
+// nodeMapV2 is an iteration on nodeMap which associates an IPSec SPI with each
+// node in the map.
+type nodeMapV2 struct {
+	conf   Config
+	bpfMap *ebpf.Map
+}
+
+func newMapV2(mapName string, conf Config) *nodeMapV2 {
+	return &nodeMapV2{
+		conf: conf,
+		bpfMap: ebpf.NewMap(&ebpf.MapSpec{
+			Name:       mapName,
+			Type:       ebpf.Hash,
+			KeySize:    uint32(unsafe.Sizeof(NodeKey{})),
+			ValueSize:  uint32(unsafe.Sizeof(NodeValueV2{})),
+			MaxEntries: conf.NodeMapMax,
+			Flags:      unix.BPF_F_NO_PREALLOC,
+			Pinning:    ebpf.PinByName,
+		}),
+	}
+}
+
+type NodeValueV2 struct {
+	NodeID uint16
+	SPI    uint8
+	Pad    uint8
+}
+
+func (m *nodeMapV2) Update(ip net.IP, nodeID uint16, SPI uint8) error {
+	key := newNodeKey(ip)
+	val := NodeValueV2{NodeID: nodeID, SPI: SPI}
+	return m.bpfMap.Update(key, val, 0)
+}
+
+func (m *nodeMapV2) Size() uint32 {
+	return m.conf.NodeMapMax
+}
+
+func (m *nodeMapV2) Delete(ip net.IP) error {
+	key := newNodeKey(ip)
+	return m.bpfMap.Map.Delete(key)
+}
+
+// NodeIterateCallback represents the signature of the callback function
+// expected by the IterateWithCallback method, which in turn is used to iterate
+// all the keys/values of a node map.
+type NodeIterateCallbackV2 func(*NodeKey, *NodeValueV2)
+
+func (m *nodeMapV2) IterateWithCallback(cb NodeIterateCallbackV2) error {
+	return m.bpfMap.IterateWithCallback(&NodeKey{}, &NodeValueV2{},
+		func(k, v interface{}) {
+			key := k.(*NodeKey)
+			value := v.(*NodeValueV2)
```

Comment:
Feels like we should be avoiding a panic just in case the typecast fails?

```suggestion
			key, ok := k.(*NodeKey)
			if !ok {
				// Some key error handling here
			}
			value, ok := v.(*NodeValueV2)
			if !ok {
				// Some value error handling here
			}
```

Reply:
Think this is total logical, will update.

Reply:
Resolved in latest push